README.md
# OraAlertCheck
(originally hosted in code.google.com/p/oraalertcheck, automatically exported here)

OraAlertCheck is a (I hope) simple korn script that checks your Oracle databases alert log files and inform you when it detect problems (ORA-NNNNN errors).

## Why Korn shell??

I choose Korn shell becouse it's considered the "standard" shell of Unix, and it's available by default on most nix systems, so no extra software is needed to be installed on your server to use OraAlertCheck.

## Installation

Checkout the files following Github instructions.

Put all the files in the same directory (example: /home/oracle/OraAlertCheck).

Rename OraAlertCheck_config.sh.sample to OraAlertCheck_config.sh. Edit the file and change the variables to suit your needs.
<b>It's important to set the WORKDIR to where the script is located</b>

OraAlertCheck looks for your instances on /etc/oratab by default, so if this does not exists you should create it or link it to the real path.

Check to see if the script can find your alert.log files running:

<pre>
$ ./OraAlertCheck.sh test
</pre>

You should see the ORACLE_SID followed by the alert.log path. If you can't and you discover a bug in the program, let me know, if it does then skip to the next paragraph. The script connects to the database "/ as sysdba" to get the value of the background_dump_dest parameter (where alert_SID.log file should be). Sometimes this parameter has characters like "?" to indicate the ORACLE_HOME. Be sure to have the full path there.

Now that the log file was found you can run the script with no parameters. Since it's the first time, it will run through all your log file and email you with a lot of errors. The script will generate a hidden file where it will set the current line number so next time it will start just where it finished.

If all worked fine, you can add the script to your crontab, every five minutes should be fine.

The script needs some work yet so please don't hesitate to send me your opinions.

NOTE: We need your help!!! In order to develop a good error detection we need every DBA using this script to send new errors that the script does not have. So please send us every error you find! You can email me to support (at) 4tm.biz, or submit your push request.

#Features

This is the list of features that currently OraAlertCheck supports:

* Parse "^ORA-NNNNN" errors and send emails depending on the severity of the error (this will be enhanced as other DBAs contribute with more and more errors).
* If the errors are on the errors list the alert will provide more information (example: if potential bug then sends a reference to the proper metalink notes).
* Errors can be costomized accordingly to db version.
* Only uses korn shell and basic commands like df, compress, tail, cat, echo, etc., so it should be very portable between nix platforms, as well as Windows using Cygwin.
* Automatic alert.log truncation
* Sends first and second parameters in ORA-00600 and ORA-07445 internal errors.


Changelog.md
# Changelog #

0.1.6
  * Updated license to GPL v3
  * Moved to Google Code for sharing source code with SVN and having a good Issue Tracker.

0.1.5

  * OraAlertCheck is fully functional in AIX (thanks Fayyaz, tested on 5.3).
  * Corrections on the way the script treats filenames. When the oracle error was ORA-00600 and had special characters on the arguments the script created temporary files with spaces or : characters which made the script failed.

0.1.4

  * OraAlertCheck is fully functional in HPUX and Solaris (thanks Raghu Maddula).
  * oratab file location can now be specified in the Config file.
  * The tail command needs to support the -n flag. In Solaris the default one did not, so a new config parameter exists to specify a new location (/usr/xpg4/bin/tail for Solaris).
  * The "test" run of the script will not only show if it find the alert.log but also will let you now if that instance won't be checked becouse the startup parameter in oratab is "N".
  * Support for detection of the "encountered when generating server alert" message was added.

0.1.3

  * Detects "caused by" and "signalled by" errors and alert them accordingly.

0.1.2

  * GNU GPL License text added to the files
  * Simpler config file lookup (just leave on the same directory as the run file)
  * Run it from any directory
  * Show the line error in the alert mails, not only the error code.

0.1.1

  * Automatic alert.log truncation
  * Send different emails if ORA-00600 arguments are differents.

0.1

  * Initial release.

OraAlertCheck_errors.sh
#!/usr/bin/ksh

##
## To be called from main file (OraAlertCheck.sh)
## Modify accordingly with your needs. Please email me
## bugs and modificactions you think it could by usefull
## (see my details in the main files)
##
## Send suggestions to newalertlogerrors@jluis.com.ar
##
##    OraAlertCheck, an Oracle alert.log watchdog
##
##    Copyright (C) 2004-2005  Jose Luis Canciani
##
##    This program is free software; you can redistribute it and/or modify
##    it under the terms of the GNU General Public License as published by
##    the Free Software Foundation; either version 2 of the License, or
##    (at your option) any later version.
##
##    This program is distributed in the hope that it will be useful,
##    but WITHOUT ANY WARRANTY; without even the implied warranty of
##    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
##    GNU General Public License for more details.
##
##    You should have received a copy of the GNU General Public License
##    along with this program; if not, write to the Free Software
##    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
##

filter_error () {
  # get error and filter according oracle version
  # $1 : oracle_sid
  # $2 : oracle version
  # $3 : ORA-NNNNN
  # $4 : optional ORA-00600/7445 first parameter NNNNNN
  # $5 : optional ORA-00600/7445 second parameter NNNNNN
  # $6 : the error line as in the alert.log
  #
  case "${3}" in
    "ORA-01110" ) return 0;; # prints datafile for previous error, we ignore it
    "ORA-01578" ) ## corruption error, send pager alert
        buildmail "${3}" "${MAIL2}" "Alerta DB ${1}" "Corruption errors! Review immediate and restore if needed" ; return 0;;
    "ORA-24756" )
        if [[ "${2}" < "8.0.6" && "${2}" != 8.0.5.1* && "${2}" != 8.0.4.3* ]]
        then
           buildmail "$3" "${MAIL1}" "Alerta DB ${1}" "NOTE: possible bug, see metalink note 62301.1" ;
           return 0;
        fi
        ;;
    #### finally, oracle internal error 600 and 7445
    "ORA-00600" )
        buildmail "${3}_${4}_${5}" "${MAIL1}" "Alerta DB ${1}" "First two arguments: [${4}] [${5}]" ; return 0;;
    "ORA-07445" )
        buildmail "${3}_${4}_${5}" "${MAIL1}" "Alerta DB ${1}" "First two arguments: [${4}] [${5}]" ; return 0;;
    * ) buildmail "$3" "${MAIL1}" "Alerta DB ${1}" "no extra OraAlertCheck information" "${6}"; return 0;;
  esac
}



trunk/OraAlertCheck.sh
#!/usr/bin/ksh

##
## Main file, call it from you crontab every 5 or 10 minutes
## Run 
##   $ OraAlertCheck test
## to see what instances and alertlog files the script detects
##
## Send suggestions to newalertlogerrors@jluis.com.ar
##
##    OraAlertCheck, an Oracle alert.log watchdog
##
##    Copyright (C) 2004-2005  Jose Luis Canciani
##
##    This program is free software; you can redistribute it and/or modify
##    it under the terms of the GNU General Public License as published by
##    the Free Software Foundation; either version 2 of the License, or
##    (at your option) any later version.
##
##    This program is distributed in the hope that it will be useful,
##    but WITHOUT ANY WARRANTY; without even the implied warranty of
##    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
##    GNU General Public License for more details.
##
##    You should have received a copy of the GNU General Public License
##    along with this program; if not, write to the Free Software
##    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
##

export SCRIPTDIR="`dirname ${0}`"

if [ -x "${SCRIPTDIR}/OraAlertCheck_config.sh" ]
then 
  . "${SCRIPTDIR}/OraAlertCheck_config.sh"
else
  echo Config file not found or cannot execute. Aborting...
  exit 1
fi

if [ ! -x "${TAIL}" ]
then
  echo "tail command not found. Edit TAIL var in config file."
  echo "  NOTE: tail command should support -n option"
fi

if [ ! -f "${ORATAB_LOC}" ]
then
  echo "oratab file not found. Edit ORATAB_LOC var in config file."
fi

sendmails () {
  ls ${WORKDIR}/.tmp.*.msg | while read EMAIL
  do
    cat "${EMAIL}" | ${SENDMAIL_COMMAND}
    rm -f "${EMAIL}"
  done
  return $?
}

buildmail () {
  # $1 ORA-NNNN
  # $2 destination mails
  # $3 subject
  # $4 extra body text
  # $5 error line as seen in alert.log
  if [ -f "${WORKDIR}/.tmp.${1}.msg" ]
  then
    COUNT=`${TAIL} -n1 "${WORKDIR}/.tmp.${1}.msg" | awk -F":" '{print $2}'`
    COUNT=`echo "${COUNT}+1" | bc`
    LINES=`wc "${WORKDIR}/.tmp.${1}.msg" | awk '{print $1}'`
    LINES=`echo "${LINES}-1" | bc`
    head -n${LINES} "${WORKDIR}/.tmp.${1}.msg" > "${WORKDIR}/.tmp.${1}.msg.wrk"
    mv "${WORKDIR}/.tmp.${1}.msg.wrk" "${WORKDIR}/.tmp.${1}.msg"
    echo "ERRORCOUNT:${COUNT}" >> "${WORKDIR}/.tmp.${1}.msg"
  else
    OERR=`echo ${1} | awk -F"_" '{print $1}' | sed s/\-/\ /`
    cat <<EOF > "${WORKDIR}/.tmp.${1}.msg"
From: ${MAILFROM}
To: $2
Subject: $3

${1} `date`
Alert log line: ${5}

oerr output:
`oerr ${OERR}`
$4
ERRORCOUNT:1
EOF
  fi
  return 0
}

# include filter error function
. "${SCRIPTDIR}/OraAlertCheck_errors.sh"

for i in `cat ${ORATAB_LOC} | grep -v "^#" | grep -v "^$" | grep -v "^*" | awk -F":" '{print $1":"$2":"$3}'`
do
  export ORACLE_SID=`echo ${i} | awk -F":" '{print $1}'`
  export ORACLE_HOME=`echo ${i} | awk -F":" '{print $2}'`
  export ORACLE_START=`echo ${i} | awk -F":" '{print $3}'`
  export PATH=${ORACLE_HOME}/bin:${PATH}
  
  export ORACLE_VERSION=`sqlplus /nolog << EOF | grep '^lookForMe' | sed 's/^lookForMe//'
conn / as sysdba
set linesize 32767
select 'lookForMe'||version
from v\\\$instance;
exit
EOF`

  export BDUMP_DIR=`sqlplus /nolog << EOF | grep '^lookForMe' | sed 's/^lookForMe//'
conn / as sysdba
set linesize 32767
select 'lookForMe'||replace(value,'?','${ORACLE_HOME}')
from v\\\$parameter
where name = 'background_dump_dest';
exit
EOF`


  if [ "$1" = "test" ]
  then
    # test mode, only prints alert log files
    if [ "${ORACLE_START}" = "Y" ]
    then
      if [ ! -f "${BDUMP_DIR}/alert_${ORACLE_SID}.log" ]
      then
        echo "Detected ${ORACLE_SID} but alertlog file NOT FOUND: ${BDUMP_DIR}/alert_${ORACLE_SID}.log"
      else
        echo "Detected ${ORACLE_SID} and alertlog file: ${BDUMP_DIR}/alert_${ORACLE_SID}.log"
      fi
    else
      echo "Detected ${ORACLE_SID} but it won't be checked (becouse oratab startup is disabled)"
    fi
  else
   if [ "${ORACLE_START}" = "Y" ]
   then

    # normal check continues

    LOG_FILE="${BDUMP_DIR}/alert_${ORACLE_SID}.log"
  
    if [ ! -f "${LOG_FILE}" ]
    then
      echo "Error: file not found ${LOG_FILE}"
      exit 1
    fi
    
    if [ ! -f "${WORKDIR}/.alert_${ORACLE_SID}.curline" ]
    then
      echo 0 > "${WORKDIR}/.alert_${ORACLE_SID}.curline"
    fi
    
    CURLINE=`cat "${WORKDIR}/.alert_${ORACLE_SID}.curline"`
    
    LINECOUNT=`wc "${LOG_FILE}" | awk '{print $1}'`
  
    if [ ${CURLINE} -gt ${LINECOUNT}  ]
    then
      # reset line count, alert log was truncated
      CURLINE=0
    fi
    
    STARTLINE=`echo "${LINECOUNT} - ${CURLINE}" | bc`
    ERRORS=0
    
    ${TAIL} -n"${STARTLINE}" "${LOG_FILE}" | grep "^ORA-" | while read LINE
    do
      ERRORS=1
      # get Oracle Error code
      echo $LINE | grep "caused by" > /dev/null
      if [ $? -eq 0 ]
      then
        ORACLE_ERROR=`echo ${LINE} | awk -F" " '{print $1}'`
      else
        echo $LINE | grep "signalled during" > /dev/null
        if [ $? -eq 0 ]
        then
          ORACLE_ERROR=`echo ${LINE} | awk '{print $1}'`
        else
          echo $LINE | grep "encountered when generating server alert" > /dev/null
          if [ $? -eq 0 ]
          then
            ORACLE_ERROR=`echo ${LINE} | awk '{print $1}'`
          else
            ORACLE_ERROR=`echo ${LINE} | awk -F":" '{print $1}'`
          fi
        fi
      fi
      # get ora-600 and ora-7445 arguments
      if [[ "${ORACLE_ERROR}" = "ORA-00600" || "${ORACLE_ERROR}" = "ORA-07445" ]]
      then
        FIRST=`echo ${LINE} | sed -e s/[\]]/\|/g | sed -e s/[\[]/\|/g | awk -F"|" '{print $2}'`
        SECOND=`echo ${LINE} | sed -e s/[\]]/\|/g | sed -e s/[\[]/\|/g | awk -F"|" '{print $4}'`
      else
        FIRST="NULL"
        SECOND="NULL"
      fi
  
      # filter the errors and build the emails
      filter_error "${ORACLE_SID}" "${ORACLE_VERSION}" "${ORACLE_ERROR}" "${FIRST}" "${SECOND}" "${LINE}"
    done
    
    echo "${LINECOUNT}" > "${WORKDIR}/.alert_${ORACLE_SID}.curline"
    
    # check to see if we should truncate the file
    if [ "${TRUNCATE_BYTES}" -gt 0 ]
    then
      CURSIZE=`ls -la "${LOG_FILE}" | awk '{print $5}'`
      if [ ${CURSIZE} -gt ${TRUNCATE_BYTES} ]
      then
        ERRORS=1
        STAMP="`date +%Y%m%d`_$$"
        
        mv "${LOG_FILE}" "${BDUMP_DIR}/alert_${ORACLE_SID}_${STAMP}.log"
        compress "${BDUMP_DIR}/alert_${ORACLE_SID}_${STAMP}.log"

	buildmail "ORA-00000" "${MAIL1}" "Notice DB ${1}" "Alert log truncated at ${CURSIZE} bytes" ;
      fi
    fi

    # send the emails
    if [ "${ERRORS}" = 1 ]
    then
      sendmails
    fi
    
   fi
  fi

done


exit 0

